src/service_daemon.rs

```diff
@@ -569,6 +625,37 @@ impl ServiceDaemon {
                 zc.add_retransmission(next_time, Command::Browse(ty, delay, listener));
             }
 
+            Command::ResolveHostname(hostname, next_delay, listener) => {
+                let addr_list: Vec<_> = zc.intf_socks.keys().collect();
+                if let Err(e) = listener.send(ResolutionEvent::SearchStarted(format!(
+                    "{} on addrs {:?}",
+                    &hostname, &addr_list
+                ))) {
+                    error!(
+                        "Failed to send ResolveStarted({})(repeating:{}): {}",
+                        &hostname, repeating, e
+                    );
+                    return;
+                }
+                if !repeating {
+                    zc.add_hostname_querier(hostname.clone(), listener.clone());
+                    // if we already have the records in our cache, just send them
+                    zc.query_cache_for_hostname(&hostname, listener.clone());
```

Comment:
I think using the cache is good. But I don't think there's going to be a lot of people running overlapping service discovery and hostname resolution threads at the same time (in the sense that they'd use the same `A`/`AAAA` records), considering the service discovery part already resolves the host for you automatically. So the value is probably miniscule.

But in the case that the host goes down inbetween the time you resolved the hostname and the time you were going to connect to the service, you'd have to handle the error while establishing a connection anyway, no matter whether you're using a 60 second old record from the cache or a record you received 20 milliiseconds beforehand. Whether a record is coming from the cache or has just been received shouldn't really matter to the user, if they're doing their error handling properly.

> If I were to resolve a hostname, I might want to always query it for the current resolution even they are already in the cache

I think it would be better to let the library user have access to a function that lets them flush either the whole, or parts of the cache. That would give them a choice to opt out, and would give them the power to do things like flushing the cache when you're having connection problems and seeing whether the host still can be resolved, to provide information about whether it's a problem with the service or the host, etc.

> How about your use case?

I'm using this for a cli tool that exits after each invocation, so the entire cache is dropped. It doesn't really affect me.